saw-core/src/SAWCore/Term/Pretty.hs

```diff
@@ -154,6 +154,40 @@ consVarNaming (VarNaming names renames used) name =
   let nm = freshName used name
   in (nm, VarNaming (nm : names) renames (Set.insert nm used))
 
+-- | Add a new variable with the given 'VarName' to the 'VarNaming',
+-- returning both the chosen fresh name and the new 'VarNaming'.
+-- As a special case, if the base name is "_", it is not modified.
+insertVarNaming :: VarNaming -> VarName -> (LocalName, VarNaming)
+insertVarNaming (VarNaming names renames used) (VarName i name) =
+  let nm = freshName used name
+  in (nm, VarNaming names (IntMap.insert i nm renames) (Set.insert nm used))
+
+-- | Compute the set of all free 'VarName's in a term.
+termVarNames :: Term -> Set VarName
+termVarNames t0 = evalState (go t0) IntMap.empty
+  where
+    go :: Term -> State (IntMap (Set VarName)) (Set VarName)
+    go tm =
+      case tm of
+        Unshared tf -> termf <$> traverse go tf
+        STApp { stAppIndex = i, stAppTermF = tf, stAppFreeVars = _vs } ->
+          do memo <- get
+             case IntMap.lookup i memo of
+               Just vars -> pure vars
+               Nothing ->
+                 do vars <- termf <$> traverse go tf
+                    modify (IntMap.insert i vars)
+                    pure vars
+    termf :: TermF (Set VarName) -> Set VarName
+    termf tf =
+      case tf of
+        FTermF ftf -> Fold.fold ftf
+        App e1 e2 -> Set.union e1 e2
+        Lambda _ e1 e2 -> Set.union e1 e2
+        Pi _ e1 e2 -> Set.union e1 e2
```

Comment:
I'm a bit surprised that the `Lambda` and `Pi` cases don't have to remove any variables that are bound by the lambda/pi abstraction itself. Am I misunderstanding what this function is doing?

Reply:
At the moment, `Lambda` and `Pi` only bind *de Bruijn* variables, not named ones. That's why they don't remove anything. When we do the de-Bruijn-to-named-variable switchover for binders, then this code will change.